core/src/main/java/io/grpc/MethodDescriptor.java

```diff
@@ -262,4 +269,40 @@ public static String extractFullServiceName(String fullMethodName) {
     }
     return fullMethodName.substring(0, index);
   }
+
+  /**
+   * An internal only class for passing information directly into the transport.
+   */
+  @Internal
+  public static final class InternalAccessor {
+    private static final String PACKAGE = InternalAccessor.class.getPackage().getName();
+
+    private static final Set<String> PACKAGE_WHITELIST =
+        new HashSet<String>(Arrays.asList(PACKAGE + ".netty."));
+
+    /**
+     * Only packages on the whitelist may access this.
+     */
+    @Internal
+    public InternalAccessor() {
+      StackTraceElement[] st = new Throwable().getStackTrace();
```

Comment:
In terms of API stability, I think having a higher barrier to entry is the key.   Marking classes as `@Internal` has worked so far, but this is the first time we are exposing something that could really break if someone started depending on it.  

I also toyed with the idea of making this hard in other ways, such as masking the return type, or giving it a weird name.  For example:

```
@Internal
public Object internalOnlyGetRawPath();
```

This would force consumers to cast it, and reduce the risk of having auto complete pick it up.  However, *our* own code would have to do the same thing, so anyone looking at it might get the idea that its okay.  With the reflect guard, there is no way to misinterpret the intent.  

This also conveniently works for internal constructors, where it is more difficult to change the name.

Thoughts?